lib/AST/ASTPrinter.cpp

```diff
@@ -2379,9 +2379,43 @@ void PrintAST::visitStructDecl(StructDecl *decl) {
   }
 }
 
+/// Determines if a class has any designated initializers that are inaccessible
+/// to classes outside this module. If that's the case, then the subclass cannot
+/// declare its own designated initializers and will not inherit the convenience
+/// initializers from the superclass.
+static bool hasInaccessibleDesignatedInit(ClassDecl *decl) {
+  // We only care for classes that show up in module interfaces.
+  if (!isPublicOrUsableFromInline(decl))
+    return false;
+
+  auto flags = OptionSet<NominalTypeDecl::LookupDirectFlags>();
+  flags |= NominalTypeDecl::LookupDirectFlags::IgnoreNewExtensions;
+  auto members = decl->lookupDirect(DeclBaseName::createConstructor(), flags);
+
+  for (auto member : members) {
+    assert(isa<ConstructorDecl>(member));
+    auto init = dyn_cast<ConstructorDecl>(member);
+    if (!init->isDesignatedInit())
+      continue;
+    if (!isPublicOrUsableFromInline(init))
```

Comment:
…and then to make this fit ASTPrinter better, you can use `shouldPrint` instead of hardcoding `isPublicOrUsableFromInline`.